#7 FilterProcessor idea

kamarton opened the pull request:
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌
| Tests pass?   | ✔️
| Fixed issues  | #7 

My idea for custom filters. Not necessarily for merging.

#### Base
- added `FilterProcessor` and `Processor` classes
- A data reader can define itself a processor interface that can use it to generate or executing filters. examples:
  - `IteratorDataReader` define processor with `->excute(..)`
  - `SqlDataReader` define processor with `->toQueryData(...)` (`return [query string, parameters]`)
- The `FilterProcessor` is accessible from the public, so anyone can add or override the processor that processes the condition. 
- `FilterProcessor` only deals with processors that it can use.

#### eg filter column in grid view
1. filter column creating own filter processors (eg PhpVariable and SqlQuery processors)
2. colum filter add the processors to girdview
3. gridview add the processors to data reader

kamarton commented:
I think this also increases the reusability of the code, e.g. if I want a different IterableDataReader then I would have to copy the filter processing at this time.

samdark commented on src/Reader/Filter/Processor/Processor.php:
```diff
@@ -0,0 +1,13 @@
+namespace Yiisoft\Data\Reader\Filter\Processor;
+
+
+interface Processor
```
The naming is confusing. Processor and FilterProcessor sound like the same thing.

kamarton replied:
> Can we name these better?
- `FilterProcessor` alternatives:
  - `FilterBuilder`
  - `FilterSet`
- `Processor` alternatives:
  - `Unit`
  - `ConditionUnit`

> Can we reduce abstraction with still achieving the same goal?

This is the first version. I think now seeing how it works can be further simplified. I'm trying to simplify it.

samdark commented on src/Reader/IterableDataReader.php:
```diff
@@ -65,50 +71,9 @@ private function sortItems(iterable $items, Sort $sort): iterable
-        }
+        $filterProcessor = $this->getFilterProcessor();
+        /* @var $filterProcessor PhpVariableFilterProcessor */
+        return $filterProcessor->execute($item, $filter);
```
To indicate that it's not building a query we can rename method:

```suggestion
        return $filterProcessor->match($item, $filter);
```

kamarton replied:
renamed

kamarton commented:
> 2. Can we reduce abstraction with still achieving the same goal?

@samdark I removed the `FilterProcessor` layer completely, and I renamed `Processor` to `Unit`.

kamarton commented:
@samdark Have an idea for filters at DataWriter? If I think well, you will need it, example. `UPDATE ... WHERE ..filters..`

samdark commented:
> Have an idea for filters at DataWriter? If I think well, you will need it, example. UPDATE ... WHERE ..filters..

That likely won't be used. I can't imagine a situation where I'd need that.

samdark commented:
> Yiisoft\Data\Reader\Filter -> Yiisoft\Data\Filter

That can be done, yes.

samdark commented on src/Reader/Filter/Unit/FilterUnitInterface.php:
```diff
@@ -0,0 +1,10 @@
+namespace Yiisoft\Data\Reader\Filter\Unit;
+
+
+interface FilterUnitInterface
```
Is "unit" something like "condition" or "criterion"?

samdark replied:
Why can't be be called "Filter" as it was before?

kamarton replied:
This was formerly called `Processor`.
- `Filter` is a criteria
- `FilterUnit` A processor that produces the actual execution of the filter, or the data needed to execute it.

samdark replied:
@yiisoft/reviewers any ideas for better naming?

samdark commented on src/Reader/Filter/Unit/VariableUnit/Any.php:
```diff
@@ -0,0 +1,24 @@
+
+    protected function checkResults(array $results): bool
+    {
+        return false;
```
Should it `foreach` here?

kamarton replied:
https://github.com/yiisoft/data/pull/20/files#diff-400ea3c48aff1d9a817c8cb53b03c9bcR11-R13

- `checkResult`: check only one result. https://github.com/yiisoft/data/pull/20/files#diff-400ea3c48aff1d9a817c8cb53b03c9bcR29

- `checkResults`: chek all result. https://github.com/yiisoft/data/pull/20/files#diff-400ea3c48aff1d9a817c8cb53b03c9bcR36

-  if criteria is end, `checkResult` exiting. https://github.com/yiisoft/data/pull/20/files#diff-400ea3c48aff1d9a817c8cb53b03c9bcR30-R32

samdark commented on src/Reader/Filter/Unit/VariableUnit/All.php:
```diff
@@ -0,0 +1,24 @@
+
+    protected function checkResults(array $results): bool
+    {
+        return true;
```
Should it `foreach` here?

samdark commented on src/Reader/FilterableDataInterface.php:
```diff
@@ -4,8 +4,10 @@
 interface FilterableDataInterface
 {
     public function withFilter(FilterInterface $filter);
+    public function withFilterUnits(FilterUnitInterface... $filterUnits);
```
That doesn't look right to me. What's the difference between Filter and FilterUnit? Which one should I use if I have `FilterableDataInterface`?

kamarton replied:
- `withFilter`: conditions for filtering
- `withFilterUnits`: the unit associated with the filter condition that performs or compiles the filter. This can be used to apply your own filters.

samdark commented on tests/Paginator/KeysetPaginatorTest.php:
```diff
@@ -77,6 +78,11 @@ public function withFilter(FilterInterface $filter)
                 // do nothing
             }
+
+            public function withFilterUnits(FilterUnitInterface ...$filterUnits)
```
How should gridview choose if `withFilter` or `withFilterUnits` should be used?

kamarton replied:
Can use both in gridview:
- `withFilter`: conditions for filtering
- `withFilterUnits`: the unit associated with the filter condition that performs or compiles the filter. This can be used to apply your own filters, and not necessarily gridview itself, but columnfilterview.

```php
$data = $data->withFilters(new RegexpFilter('field', '^apple-\d+$'))
  ->withFilterUnits(
    new RegexpFilterVariable(),
    new RegexpFilterSql(),
    new RegexpFilterMongoDb(),
    new RegexpFilterFacebookMessageApi()
  );
```

samdark replied:
Aha! So "filter" is condition and "filter unit" is what applies condition?

samdark replied:
"filter unit" is data reader specific, right?

kamarton replied:
Data reader and writer specific. Example

```sql
Select ... From ... Where <filters>
Update ... Set ... Where <filters>
```

kamarton replied:
> Aha! So "filter" is condition and "filter unit" is what applies condition?

Yes

samdark commented on src/Reader/Filter/Unit/VariableUnit/Equals.php:
```diff
@@ -0,0 +1,22 @@
+
+use Yiisoft\Data\Reader\Filter\Unit\FilterUnitInterface;
+
+class Equals implements VariableUnitInterface, FilterUnitInterface
```
I can't find implementation of `VariableUnitInterface` anywhere :(

kamarton replied:
https://github.com/yiisoft/data/pull/20/files#diff-0139522285dbe44fc135d09daf4e9f2bR1-R9

This is a interface for PHP variable match. Most interfaces need to be defined for SQL, mongodb and others.

samdark commented:
@kamarton I've adjusted naming. Looks better regarding understanding what is what to me.

I'm still concerned with namespaces:

1. `FilterInterface` and `FilterProcessorInterface` look like general things to me so could be moved to namespace `Yiisoft\Data\Reader`.
2. `Yiisoft\Data\Reader\Filter\Processor\Iterable` is closely related with `Yiisoft\Data\Reader\IterableDataReader`.

So it makes sense to me to have the following structure:

```
Yiisoft\Data\Reader\
    Filter\ -- filters here
    Iterable\
        Processor\ -- iterable-specific processors here
        IterableDataReader
    FilterInterface
    FilterProcessorInterface    
```

What do you think?

kamarton commented:
I like it, but depending on the #22 issue, it may not be applicable. The `iIterable` namespace is a good idea, and I would move the` Sort` class there as well.

```
Yiisoft\Data\Reader\
    Filter\ -- filters here
    Iterable\
        Processor\ -- iterable-specific processors here
        IterableDataReader
        Sort                  -- This also belongs to the iterator. (#22)
    FilterInterface           -- currently like it. (#22)
    FilterProcessorInterface  -- currently like it. (#22)
```

samdark commented:
I don't think #22 is valid.

kamarton commented:
I was thinking a little and it might be more logical to put it under the `Filter` namespace.

```php
Yiisoft\Data\Reader\
    Filter\ -- filters here
        FilterInterface           <----- (Currently located here.)
        FilterProcessorInterface  <-----
    Iterable\
        Processor\ -- iterable-specific processors here
        IterableDataReader
```

kamarton commented:
![Desktop screenshot](https://user-images.githubusercontent.com/9432032/65298854-57a72200-db6d-11e9-96b6-04cb27d073a9.png)

samdark commented:
I prefer keeping interfaces one level above implementations (if there's more than a single implementation) but what you have is good.

samdark commented:
One extra thing that could be improved is extra docs on how to add your own filter and filter processor. If you have extra time, please do a draft. If not, I'll add it before merging.

kamarton commented:
@samdark You mean [readme.md](https://github.com/yiisoft/data/blob/master/README.md) or somewhere else?

samdark commented:
Yes. There.

kamarton commented:
Document added

samdark commented:
@kamarton would you please merge master into this PR branch?

kamarton commented:
Merged from upstream/master. 

Please check my edits of documentation before merging to master (my English is poor).

samdark commented:
Merged. Thank you!
